sdk/webpubsub/azure-messaging-webpubsub-client/src/main/java/com/azure/messaging/webpubsub/client/WebPubSubAsyncClient.java

```diff
@@ -0,0 +1,952 @@
+    private final AtomicReference<Disposable> sequenceAckTask = new AtomicReference<>();
+
+    // client state
+    private final ClientState clientState = new ClientState();
+    // state on close
+    private final AtomicBoolean isDisposed = new AtomicBoolean();
+    private final Sinks.Empty<Void> isClosedMono = Sinks.empty();
+    // state on stop by user
+    private final AtomicBoolean isStoppedByUser = new AtomicBoolean();
+    private final AtomicReference<Sinks.Empty<Void>> isStoppedByUserMono = new AtomicReference<>();
+
+    // groups
+    private final ConcurrentMap<String, WebPubSubGroup> groups = new ConcurrentHashMap<>();
+
+    // retry
+    private final Retry sendMessageRetrySpec;
+
+    private static final Duration ACK_TIMEOUT = Duration.ofSeconds(30);
+    private static final Duration RECOVER_TIMEOUT = Duration.ofSeconds(30);
+    private static final Retry RECONNECT_RETRY_SPEC =
+        Retry.backoff(Long.MAX_VALUE, Duration.ofSeconds(1))
+            .filter(thr -> !(thr instanceof StopReconnectException));
+
+    WebPubSubAsyncClient(Mono<String> clientAccessUriProvider,
+                         WebPubSubProtocol webPubSubProtocol,
+                         RetryStrategy retryStrategy,
+                         boolean autoReconnect,
+                         boolean autoRestoreGroup) {
+
+        this.logger = new ClientLogger(WebPubSubAsyncClient.class);
+
+        this.clientAccessUriProvider = Objects.requireNonNull(clientAccessUriProvider);
+        this.webPubSubProtocol = Objects.requireNonNull(webPubSubProtocol);
+        this.autoReconnect = autoReconnect;
+        this.autoRestoreGroup = autoRestoreGroup;
+
+        this.clientManager = ClientManager.createClient();
+
+        Objects.requireNonNull(retryStrategy);
+        this.sendMessageRetrySpec = Retry.from(signals -> {
+            AtomicInteger retryCount = new AtomicInteger(0);
+            return signals.concatMap(s -> {
+                Mono<Retry.RetrySignal> ret = Mono.error(s.failure());
+                if (s.failure() instanceof SendMessageFailedException) {
+                    if (((SendMessageFailedException) s.failure()).isTransient()) {
+                        int retryAttempt = retryCount.incrementAndGet();
+                        if (retryAttempt <= retryStrategy.getMaxRetries()) {
+                            ret = Mono.delay(retryStrategy.calculateRetryDelay(retryAttempt))
+                                .then(Mono.just(s));
+                        }
+                    }
+                }
+                return ret;
+            });
+        });
+    }
+
+    /**
+     * Gets the connection ID.
+     *
+     * @return the connection ID.
+     */
+    public String getConnectionId() {
+        return connectionId;
+    }
+
+    /**
+     * Starts the client for connecting to the server.
+     *
+     * @return the task.
+     */
+    public Mono<Void> start() {
+        if (clientState.get() != WebPubSubClientState.STOPPED) {
```

Comment:
Currently `close` is modeled as Java resource https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

This design is not final.

Reply:
I was referencing the EH/SB processor client, it can re-start() after stop()/close(). So I thinking about do we need to add the ability that we can restart client after it was closed.

But WebPubSubAsyncClient is a bit different, the send/receive function need user to create the subscriber, and the close() will complete them all, so if we need restart the client, user need to re-subscribe to the new "message/event sink".

Consider the workload, I think currently design is fine. The error already tell start() can only be called when client is stopped. I am ok with it now.

Reply:
As discussed, EHSB is somewhat some same.

Their start/stop is more like resume (and start if not started)/pause. And their close like stop here.

I am not sure whether webpubsub can do this pause/resume. I guess it will just lose all the messages during the pause.

Hence the API is not directly comparable by name.